File: src/main/scala/com/pygostylia/osprey/MinecraftOutputStream.scala
package com.pygostylia.osprey

import java.io.{ByteArrayOutputStream, DataOutputStream, OutputStream}
import java.util.UUID

class MinecraftOutputStream(os: OutputStream) extends DataOutputStream(os) {
  def this() = this(new ByteArrayOutputStream())

  def writeVarInt(v: Int): Unit = {
    var i = v
    while (i > 0x7f || i < 0) {
      write(i & 0x7f | 0x80)
      i >>>= 7
    }
    write(i)
  }

  def writeString(s: String): Unit = {
    writeVarInt(s.length)
    writeBytes(s)
  }

  def writeBytes(bs: Array[Byte]): Unit = write(bs)

  def writeUUID(uuid: UUID): Unit = {
    writeLong(uuid.getMostSignificantBits)
    writeLong(uuid.getLeastSignificantBits)
  }

  def writeLocation(location: Location): Unit = {
    writeLong(location.encode)
  }

  def writeLocation(x: Int, y: Int, z: Int): Unit = {
    writeLocation(new Location(x, y, z))
  }

  def writePosition(position: Position): Unit = {
    writeDouble(position.x)
    writeDouble(position.y)
    writeDouble(position.z)
    writeByte(position.pitchAngle)
    writeByte(position.yawAngle)
  }

  def toByteArray: Array[Byte] = out.asInstanceOf[ByteArrayOutputStream].toByteArray
}

File: src/main/java/com/pygostylia/osprey/MinecraftOutputStream.java
package com.pygostylia.osprey;

import java.io.*;
import java.util.UUID;

public class MinecraftOutputStream extends DataOutputStream {
    public MinecraftOutputStream(OutputStream out) {
        super(out);
    }

    public MinecraftOutputStream() {
        super(new ByteArrayOutputStream());
    }

    public void writeString(String str) throws IOException {
        VarInt.write(this, str.length());
        writeBytes(str);
    }

    public void writeString0(String str) throws IOException {
        writeBytes(str);
        write(0);
    }

    public void writeVarInt(int number) throws IOException {
        VarInt.write(this, number);
    }

    public void writeByte(byte b) throws IOException {
        write(b);
    }

    public void writeUUID(UUID uuid) throws IOException {
        writeLong(uuid.getMostSignificantBits());
        writeLong(uuid.getLeastSignificantBits());
    }

    public void writeLocation(int x, int y, int z) throws IOException {
        long encoded = (((long) x & 0x3FFFFFF) << 38) | (((long) z & 0x3FFFFFF) << 12) | (y & 0xFFF);
        writeLong(encoded);
    }

    public void writeLocation(BlockPosition blockPosition) throws IOException {
        writeLocation(blockPosition.x(), blockPosition.y(), blockPosition.z());
    }

    public void writePosition(EntityPosition entityPosition) throws IOException {
        writeDouble(entityPosition.x);
        writeDouble(entityPosition.y);
        writeDouble(entityPosition.z);
        writeByte(entityPosition.pitchAngle());
        writeByte(entityPosition.yawAngle());
    }

    public byte[] toByteArray() {
        if (out instanceof ByteArrayOutputStream bout) {
            return bout.toByteArray();
        } else {
            return null;
        }
    }
}
